src/MongoDB.Driver.Core/Core/Bindings/CoreServerSessionPool.cs

```diff
@@ -24,8 +23,7 @@ internal class CoreServerSessionPool : ICoreServerSessionPool
     {
         // private fields
         private readonly ICluster _cluster;
-        private readonly object _lock = new object();
-        private readonly List<ICoreServerSession> _pool = new List<ICoreServerSession>();
+        private readonly LifoPool<ICoreServerSession> _pool = new LifoPool<ICoreServerSession>();
```

Comment:
The new LifoPool class holds the shared logic between the session and connection pools.